### PR TITLE
feat: throw exception if reads corrupt data

### PR DESCRIPTION
**Pull request by levy5307**
In previous implementation, server will coredump if it reads corrupt data. 
In this pull request, I changed the way to throw exception. So someone can catch the exception if he don't want to make server crash.

### CODE REVIEW

File: src/replica/replica_2pc.cpp [hunk 1]
```diff
@@ -76,6 +76,15 @@ void replica::on_client_write(dsn::message_ex *request, bool ignore_throttling)
     }
 
     task_spec *spec = task_spec::get(request->rpc_code());
+    if (dsn_unlikely(nullptr == spec || request->rpc_code() == TASK_CODE_INVALID)) {
+        dwarn_f("recv message with unhandled rpc name {} from {}, trace_id = {}",
```
Comment (hycdong) [inline src/replica/replica_2pc.cpp:80]: How about updating warning level into error level?

**Comment by hycdong** [comment]
I think `feat` may be better than `refactor` to describe this pull request.

File: src/replica/replica_2pc.cpp (hunk 1 above)
Comment (levy5307) [reply src/replica/replica_2pc.cpp:80]: OK

**Review by hycdong — APPROVED** [review]

**Review by Shuo-Jia — APPROVED** [review]

File: include/dsn/utility/binary_reader.h [hunk 2]
```diff
@@ -39,21 +40,27 @@ class binary_reader
     int read(/*out*/ bool &val) { return read_pod(val); }
 
     int read(/*out*/ std::string &s);
-    int read(char *buffer, int sz);
+    virtual int read(char *buffer, int sz);
```
Comment (Smityz) [inline include/dsn/utility/binary_reader.h:43]: It's an important path on write, add a virtual function will add a performance cost in c++, is this change necessary？

Comment (levy5307) [reply include/dsn/utility/binary_reader.h:43]: The performance cost introduced by vtable can be negligible. And this pull request is used for dealing with dirty mutation log.
